src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationSimulator.cs

```diff
@@ -94,6 +99,20 @@ public class UserOperationSimulator : IUserOperationSimulator
             UInt256? timestamp = null,
             CancellationToken cancellationToken = default)
         {
+            if (userOperation.AlreadySimulated)
+            {
+                // codehash of all accessed addresses should not change between calls
+                foreach (KeyValuePair<Address, Keccak> kv in userOperation.AddressesToCodeHashes)
+                {
+                    (Address address, Keccak expectedCodeHash) = kv;
+                    Keccak codeHash = _stateProvider.GetCodeHash(address);
```

Comment:
Is this access thread-safe?

Reply:
i think so, the initial simulation sets the addressesToCodeHashes and we check on the second simulation. but the second simulation can't happen until the second one does